src/eddymotion/model/dipy.py

```diff
@@ -406,14 +436,173 @@ def compute_pairwise_angles(
 
     """
 
-    bvecs_X = gtab_X.bvecs.T
-    bvecs_X = np.array(bvecs_X) / np.linalg.norm(bvecs_X, axis=0)
+    bvecs_X = getattr(gtab_X, "bvecs", gtab_X)
+    bvecs_X = np.array(bvecs_X.T) / np.linalg.norm(bvecs_X, axis=1)
 
     if gtab_Y is None:
         bvecs_Y = bvecs_X
     else:
-        bvecs_Y = gtab_Y.bvecs.T
-        bvecs_Y = np.array(bvecs_Y) / np.linalg.norm(bvecs_Y, axis=0)
+        bvecs_Y = np.array(getattr(gtab_Y, "bvecs", gtab_Y))
+        if bvecs_Y.ndim == 1:
+            bvecs_Y = bvecs_Y[np.newaxis, ...]
+        bvecs_Y = bvecs_Y.T / np.linalg.norm(bvecs_Y, axis=1)
 
     cosines = np.clip(bvecs_X.T @ bvecs_Y, -1.0, 1.0)
     return np.arccos(np.abs(cosines) if closest_polarity else cosines)
+
+
+class PairwiseOrientationKernel(Kernel):
+    """A scikit-learn's kernel for DWI signals."""
+
+    def __init__(
+        self,
+        weighting: str = "exponential",
+        lambda_s: float = 2.0,
+        a: float = 0.1,
+        sigma_sq: float = 1.0,
+        lambda_s_bounds: tuple[float, float] = (1e-5, 1e4),
+        a_bounds: tuple[float, float] = (1e-5, np.pi),
+        sigma_sq_bounds: tuple[float, float] = (1e-5, 1e4),
+    ):
+        self._weighting = weighting  # For __repr__
+        self.lambda_s = lambda_s
+        self.a = a
+        self.sigma_sq = sigma_sq
+        self.lambda_s_bounds = lambda_s_bounds
+        self.a_bounds = a_bounds
+        self.sigma_sq_bounds = sigma_sq_bounds
+
+    @property
+    def hyperparameter_lambda_s(self):
+        return Hyperparameter("lambda_s", "numeric", self.lambda_s_bounds)
+
+    @property
+    def hyperparameter_a(self):
+        return Hyperparameter("a", "numeric", self.a_bounds)
+
+    @property
+    def hyperparameter_sigma_sq(self):
+        return Hyperparameter("sigma_sq", "numeric", self.sigma_sq_bounds)
+
+    def __call__(self, gtab_X, gtab_Y=None, eval_gradient=False):
+        """
+        Return the kernel K(gtab_X, gtab_Y) and optionally its gradient.
+
+        Parameters
+        ----------
+        gtab_X: :obj:`~dipy.core.gradients.GradientTable`
+            Gradient table (X)
+        gtab_Y: :obj:`~dipy.core.gradients.GradientTable`
+            Gradient table (Y, optional)
```

Comment:
gtab_Y: :obj:\`~dipy.core.gradients.GradientTable\`, optional

Wondering if you left (X) and (Y, optional) so as to identify them better with scikit-learn.

Reply:
> Wondering if you left (X) and (Y, optional) so as to identify them better with scikit-learn.

Yes